### ISSUE TITLE
Handle external dependencies without forcing the user to manually download everything

### ISSUE THREAD

**Issue by cjdb** [cjdb/doge#7]
The current project requires programmers to manually download all external dependencies. This is unacceptable for a useful project.

**Comment by mmha** [comment]
I don't think you should use CMake for that. Have you considered using [Conan](https;//conan.io) instead?
The Bincrafters community already packaged [glfw](https://bintray.com/bincrafters/public-conan/glfw%3Abincrafters). In the conan-community repository there is the most recent [Boost](https://bintray.com/conan-community/conan/boost%3Aconan#), leaving only cmcstl2 and glm to be packaged. They are header-only, that should make it trivial.

It should also be possible to write a profile that automatically downloads and compiles gcc, CMake and ninja to lower the entry barrier for people to hack on the code. [CMake](https://bintray.com/conan-community/conan/cmake_installer%3Aconan) and [MinGW](https://bintray.com/conan-community/conan/mingw_installer%3Aconan) are already available and for ninja I already have an unpublished installer I could submit.

**Comment by mmha** [comment]
I did a more thorough investigation of all libraries used and available Conan packages:

| Package | State | Proposed Package Reference |
|----------|------|------------------------------|
| glm | [Already packaged](https://bintray.com/dimi309/conan-packages/glm%3Ag-truc) | glm/0.9.8.5@g-truc/stable |
| catch2 | [Already packaged](https://bintray.com/bincrafters/public-conan/catch2%3Abincrafters) | catch2/2.1.2@bincrafters/stable |
| GSL | [Already packaged](https://bintray.com/bincrafters/public-conan/gsl_microsoft%3Abincrafters) | gsl_microsoft/20180102@bincrafters/stable |
| GLFW | [Already packaged](https://bintray.com/bincrafters/public-conan/glfw%3Abincrafters) | glfw/3.2.1@bincrafters/stable |
| cmcstl2 | Created PR: CaseyCarter/cmcstl2#137 | currently cmcstl2/5e8577e@mmha/stable, but should be changed once merged
| stb | Tracked by issue: nothings/stb#566, apparently WIP: nothings/stb#555 | N/A |
| gli | Not packaged yet | N/A |

We should probably write a Conan recipe for gli, try to upstream it and wait until the work on the stb package is done. In the meantime, both could be just copied into the repository as a workaround.

When all dependencies are packaged, it is probably best to create a new Bintray repository and link all packages in. This simplifies the users work flow because they only need to add one remote at most and the README could be extended like this:

### Getting Started
1. Install gcc >= 6.3, CMake >= 3.10, Conan >= 1.0. On Ubuntu, this works like this:
```bash
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt update
sudo apt install gcc-8 python3-pip
pip3 --user install conan cmake
```
2. Add the DOGE repository to your Conan remotes:
```bash
conan remote add doge https://api.bintray.com/conan/doge/conan
```
3.  Add this to your CMakeLists:
```cmake
    conan_cmake_run(REQUIRES doge/0.42@doge/stable
                    BASIC_SETUP CMAKE_TARGETS
                    BUILD missing)
    find_package(DOGE REQUIRED)
    target_link_libraries(app PRIVATE DOGE::doge)
```

**Comment by cjdb** [comment]
Thanks for this. How easy is it to _update_ a package? I am concerned about doge being available upstream while it is pre-alpha.

How will this affect the existing CMake files?

> We should probably write a Conan recipe for gli

I have no intention to move to GLI at present; this is a dead dependency for now, and is best to focus on getting stb into Conan instead. GLI might return in the future, but there's nothing in the pipeline right now.

> Install gcc >= 6.3

Is this a generic requirement for Conan? doge exclusively supports GCC 7 at present, and will be abandoning GCC 7 as soon as GCC 8 is released.

**Comment by mmha** [comment]
> How easy is it to update a package? 

It's a matter of changing the package reference to a different version. Semver and version ranges are supported if we tag our versions appropriately and users can update via `conan install -u`.

> I am concerned about doge being available upstream while it is pre-alpha.

Conan offers different release channels in addition to the version. For example, you can choose to instruct Travis to package up different branches into channels of the same name. The invocation line for conan for that could be `conan create .. doge/$(git rev-parse --abbrev-ref HEAD)`. So users could write `doge/1.*@doge/stable` and won't be affected by updates to `master`.

You may also choose not to package doge at all and just use it for retrieving your dependencies for now.

> Is this a generic requirement for Conan? doge exclusively supports GCC 7 at present, and will be abandoning GCC 7 as soon as GCC 8 is released.

This was just an example. No, Conan makes no assumptions about which compiler you use, but it has a sanity check for supported compilers listed in `~/.conan/settings.yml`. If you use gcc 8, you may need to add that version to the list of supported versions in there until Conan is updated. But this is documented and expected behavior.
